src/Microsoft.AspNetCore.Hosting/Internal/HostingApplication.cs

```diff
@@ -64,21 +74,65 @@ public Context CreateContext(IFeatureCollection contextFeatures)
                 // Non-inline
                 LogRequestStarting(httpContext);
             }
-
             if (diagnoticsEnabled)
             {
+#if NETSTANDARD2_0
+                context.Activity = StartActivity(httpContext);
+#endif
                 // Non-inline
                 RecordBeginRequestDiagnostics(httpContext, startTimestamp);
             }
 
-            // Create and return the request Context
-            return new Context
+            // Fill and return the request Context
+            context.HttpContext = httpContext;
+            context.Scope = scope;
+            context.StartTimestamp = startTimestamp;
+            return context;
+        }
+
+#if NETSTANDARD2_0
+        private Activity StartActivity(HttpContext httpContext)
+        {
+            var activity = new Activity(ActivityName);
+            StringValues requestId;
+            if (httpContext.Request.Headers.TryGetValue(_activityTrackingOptions.RequestIdHeaderName, out requestId))
             {
-                HttpContext = httpContext,
-                Scope = scope,
-                StartTimestamp = startTimestamp,
-            };
+                try
+                {
+                    activity.SetParentId(requestId);
+                }
+                catch (ArgumentException ex)
+                {
+                    _logger.LogWarning(0, ex, "Request ID is invalid '{RequestId}", requestId);
+                }
+
+                // We expect baggage to be empty by default
+                // Only very advanced users will be using it in near future, we encouradge them to keep baggage small (few items)
+                string[] baggage = httpContext.Request.Headers.GetCommaSeparatedValues(_activityTrackingOptions.BaggageHeaderName);
+                if (baggage != StringValues.Empty)
+                {
+                    foreach (var item in baggage)
+                    {
+                        NameValueHeaderValue baggageItem;
+                        if (NameValueHeaderValue.TryParse(item, out baggageItem))
+                        {
+                            try
+                            {
+                                activity.AddBaggage(baggageItem.Name, baggageItem.Value);
+                            }
+                            catch (ArgumentException ex)
+                            {
+                                _logger.LogWarning(0, ex, "Invalid baggage item '{ItemName}' with value '{ItemValue}'", baggageItem.Name, baggageItem.Value);
+                            }
+                        }
+                    }
+                }
+            }
+
```

Comment:
Creating activity and reading headers is already performance impact does not running `Start()` actually save us from anything?